## SF Crime Kaggle Competition

The Kaggle competition, [San Francisco Crime Classification](https://www.kaggle.com/c/sf-crime), has given 11 years of incidents reported to the SFD Crime Incident Reporting. The goal is being able to predict the category of the crime with the information given to me in the dataset. Currently the highest score is 2.29303.

#### My Goals
I want to use this dataset to get an idea of what submitting a model to a kaggle compeittion looks like and also get a feeling for how far basic data analysis will get you in a competition. In terms of skill building I want to be able to optomize my models and create some visualizations.


#### Data Exploration
As usual lets start with some basic data exploration.

In [1]:
import pandas as pd
df = pd.read_csv('./data/train.csv')
df

Dates                Category  \
0       2015-05-13 23:53:00                WARRANTS   
1       2015-05-13 23:53:00          OTHER OFFENSES   
2       2015-05-13 23:33:00          OTHER OFFENSES   
3       2015-05-13 23:30:00           LARCENY/THEFT   
4       2015-05-13 23:30:00           LARCENY/THEFT   
5       2015-05-13 23:30:00           LARCENY/THEFT   
6       2015-05-13 23:30:00           VEHICLE THEFT   
7       2015-05-13 23:30:00           VEHICLE THEFT   
8       2015-05-13 23:00:00           LARCENY/THEFT   
9       2015-05-13 23:00:00           LARCENY/THEFT   
10      2015-05-13 22:58:00           LARCENY/THEFT   
11      2015-05-13 22:30:00          OTHER OFFENSES   
12      2015-05-13 22:30:00               VANDALISM   
13      2015-05-13 22:06:00           LARCENY/THEFT   
14      2015-05-13 22:00:00            NON-CRIMINAL   
15      2015-05-13 22:00:00            NON-CRIMINAL   
16      2015-05-13 22:00:00                 ROBBERY   
17      2015-05-13 21:55:00                 ASSAULT   
18      2015-05-13 21:40:00          OTHER OFFENSES   
19      2015-05-13 21:30:00            NON-CRIMINAL   
20      2015-05-13 21:30:00           LARCENY/THEFT   
21      2015-05-13 21:17:00                 ROBBERY   
22      2015-05-13 21:11:00                WARRANTS   
23      2015-05-13 21:11:00            NON-CRIMINAL   
24      2015-05-13 21:10:00           LARCENY/THEFT   
25      2015-05-13 21:00:00            NON-CRIMINAL   
26      2015-05-13 21:00:00           LARCENY/THEFT   
27      2015-05-13 21:00:00           LARCENY/THEFT   
28      2015-05-13 21:00:00           LARCENY/THEFT   
29      2015-05-13 20:56:00          OTHER OFFENSES   
...                     ...                     ...   
878019  2003-01-06 02:37:00          OTHER OFFENSES   
878020  2003-01-06 02:32:00          OTHER OFFENSES   
878021  2003-01-06 02:24:00               VANDALISM   
878022  2003-01-06 02:16:00           VEHICLE THEFT   
878023  2003-01-06 02:15:00           LARCENY/THEFT   
878024  2003-01-06 02:09:00          OTHER OFFENSES   
878025  2003-01-06 02:06:00          OTHER OFFENSES   
878026  2003-01-06 02:06:00                WARRANTS   
878027  2003-01-06 02:00:00                WARRANTS   
878028  2003-01-06 02:00:00                 ASSAULT   
878029  2003-01-06 01:54:00          OTHER OFFENSES   
878030  2003-01-06 01:54:00   SEX OFFENSES FORCIBLE   
878031  2003-01-06 01:50:00                 ASSAULT   
878032  2003-01-06 01:36:00          OTHER OFFENSES   
878033  2003-01-06 01:30:00               VANDALISM   
878034  2003-01-06 01:30:00                TRESPASS   
878035  2003-01-06 00:55:00                 ASSAULT   
878036  2003-01-06 00:55:00           LARCENY/THEFT   
878037  2003-01-06 00:55:00               VANDALISM   
878038  2003-01-06 00:42:00                WARRANTS   
878039  2003-01-06 00:40:00          OTHER OFFENSES   
878040  2003-01-06 00:33:00                 ASSAULT   
878041  2003-01-06 00:31:00          OTHER OFFENSES   
878042  2003-01-06 00:20:00                 ASSAULT   
878043  2003-01-06 00:20:00          OTHER OFFENSES   
878044  2003-01-06 00:15:00                 ROBBERY   
878045  2003-01-06 00:01:00           LARCENY/THEFT   
878046  2003-01-06 00:01:00           LARCENY/THEFT   
878047  2003-01-06 00:01:00               VANDALISM   
878048  2003-01-06 00:01:00  FORGERY/COUNTERFEITING   

                                         Descript  DayOfWeek  PdDistrict  \
0                                  WARRANT ARREST  Wednesday    NORTHERN   
1                        TRAFFIC VIOLATION ARREST  Wednesday    NORTHERN   
2                        TRAFFIC VIOLATION ARREST  Wednesday    NORTHERN   
3                    GRAND THEFT FROM LOCKED AUTO  Wednesday    NORTHERN   
4                    GRAND THEFT FROM LOCKED AUTO  Wednesday        PARK   
5                  GRAND THEFT FROM UNLOCKED AUTO  Wednesday   INGLESIDE   
6                               STOLEN AUTOMOBILE  Wednesday   INGLESIDE   
7  

This time around the data is categorical but there are only 8 independent variables and 1 dependent variable. My instinct tells me to still try Random Forest Classification but lets continue to explore the data a little bit.

In [2]:
# Here I've created a dataFrame that groups the Categories of crime and then the Addresses
# This returns the number of times a certain crime happened at each address.
crime_address = pd.DataFrame({'count': df.select_dtypes(include=['O']).groupby(['Category','Address']).size()}).reset_index()
crime_address

Category                                  Address  count
0             ARSON                       0 Block of 12TH ST      2
1             ARSON                       0 Block of 14TH ST      1
2             ARSON                        0 Block of 3RD ST      2
3             ARSON                        0 Block of 4TH ST      1
4             ARSON                        0 Block of 6TH ST     10
5             ARSON                        0 Block of 9TH ST      2
6             ARSON                   0 Block of ARKANSAS ST      1
7             ARSON                    0 Block of BALDWIN CT      1
8             ARSON                    0 Block of BAYVIEW ST      1
9             ARSON                   0 Block of BEATRICE LN      1
10            ARSON                      0 Block of BEMIS ST      1
11            ARSON                  0 Block of BERKSHIRE WY      1
12            ARSON                     0 Block of BERTHA LN      1
13            ARSON                  0 Block of BLYTHDALE AV      1
14            ARSON                   0 Block of BOUTWELL ST      1
15            ARSON                   0 Block of BRIGHTON AV      1
16            ARSON                      0 Block of BROOK ST      1
17            ARSON                  0 Block of BROOKDALE AV      3
18            ARSON                      0 Block of BRUCE AV      1
19            ARSON                     0 Block of BYXBEE ST      1
20            ARSON                      0 Block of CAINE AV      1
21            ARSON                    0 Block of CAMERON WY      3
22            ARSON                    0 Block of CAMPTON PL      1
23            ARSON                   0 Block of CASHMERE ST      2
24            ARSON                      0 Block of CEDAR ST      1
25            ARSON                    0 Block of CHENERY ST      1
26            ARSON                  0 Block of CHUMASERO DR      2
27            ARSON                   0 Block of CLARENCE PL      1
28            ARSON                       0 Block of CLAY ST      1
29            ARSON                     0 Block of COMMER CT      1
...             ...                                      ...    ...
170703  WEAPON LAWS                 VERMONT ST / DIVISION ST      1
170704  WEAPON LAWS                 VERMONT ST / MARIPOSA ST      1
170705  WEAPON LAWS                     VICENTE ST / 33RD AV      1
170706  WEAPON LAWS                     VIDAL DR / GARCES DR      1
170707  WEAPON LAWS                  VISITACION AV / CORA ST      1
170708  WEAPON LAWS                   VISITACION AV / REY ST      1
170709  WEAPON LAWS               VISITACION AV / RUTLAND ST      1
170710  WEAPON LAWS                      WALLER ST / COLE ST      1
170711  WEAPON LAWS                   WALLER ST / SHRADER ST      1
170712  WEAPON LAWS                   WALLER ST / STANYAN ST      2
170713  WEAPON LAWS                      WARD ST / GIRARD ST      1
170714  WEAPON LAWS                 WASHINGTON ST / DRUMM ST      1
170715  WEAPON LAWS  WASHINGTON ST / THE EMBARCADERONORTH ST      1
170716  WEAPON LAWS                      WAWONA ST / 21ST AV      1
170717  WEAPON LAWS                      WAWONA ST / 46TH AV      1
170718  WEAPON LAWS                   WEBSTER ST / FULTON ST      1
170719  WEAPON LAWS                    WEBSTER ST / GEARY BL      1
170720  WEAPON LAWS                    WEBSTER ST / HAYES ST      1
170721  WEAPON LAWS                  WEBSTER ST / HERMANN ST      1
170722  WEAPON LAWS                      WEBSTER ST / OAK ST      1
170723  WEAPON LAWS                 WESTPORTAL AV / ULLOA ST      1
170724  WEAPON LAWS             WHITNEY YOUNG CR / MABREY CT      1
170725  WEAPON LAWS                       WIESE ST / 15TH ST      1
170726  WEAPON LAWS                     WILLIAMS AV / 3RD ST      2
170727  WEAPON LAWS                 WILLIAMS AV / NEWHALL ST      1
170728  WEAPON LAWS                   WILLOW ST / VANNESS AV      1
170729  WEAPON LAWS                    WOOL ST / CORTLAND AV      1

In [3]:
# This reduces the information to show just the area with the highest frequency for each crime.
address_count = crime_address.groupby(['Category']).apply(lambda x: x[x['count']==x['count'].max()]).reset_index(1)

# Cleaning up the data, groupby doesn't always comeout pretty
address_count.drop(address_count.columns[[0, 1]], axis=1, inplace=True)
address_count = address_count.reset_index()

# I don't want to see rare crimes with multiple addresses tied for first with 1 occurance
address_count = address_count[address_count['count'] > 1]
address_count

Category                         Address  count
0                         ARSON          800 Block of BRYANT ST     41
1                       ASSAULT          800 Block of BRYANT ST   1926
2                    BAD CHECKS          800 Block of BRYANT ST     12
3                       BRIBERY          800 Block of BRYANT ST     12
4                      BURGLARY          800 Block of BRYANT ST    384
5            DISORDERLY CONDUCT        1000 Block of POTRERO AV    104
6   DRIVING UNDER THE INFLUENCE          800 Block of BRYANT ST     41
7                 DRUG/NARCOTIC        2000 Block of MISSION ST   1866
8                   DRUNKENNESS          800 Block of BRYANT ST    100
9                  EMBEZZLEMENT          800 Block of MARKET ST     42
10                    EXTORTION          800 Block of BRYANT ST     13
11              FAMILY OFFENSES        1000 Block of POTRERO AV     91
12       FORGERY/COUNTERFEITING          800 Block of BRYANT ST    614
13                        FRAUD          800 Block of BRYANT ST    942
14                     GAMBLING          700 Block of KEARNY ST      8
15                   KIDNAPPING          800 Block of BRYANT ST     97
16                LARCENY/THEFT          800 Block of BRYANT ST   6144
17                  LIQUOR LAWS             800 Block of 3RD ST     51
18                    LOITERING  MISSION ST / SOUTH VAN NESS AV     49
19               MISSING PERSON         1400 Block of PHELPS ST   1468
20                 NON-CRIMINAL          800 Block of BRYANT ST   5583
21               OTHER OFFENSES          800 Block of BRYANT ST   3019
44                 PROSTITUTION              ELLIS ST / HYDE ST    422
45            RECOVERED VEHICLE    200 Block of INTERSTATE80 HY     12
46                      ROBBERY          800 Block of BRYANT ST    591
47                      RUNAWAY           1200 Block of PAGE ST    338
48              SECONDARY CODES          800 Block of BRYANT ST    312
49        SEX OFFENSES FORCIBLE          800 Block of BRYANT ST    296
50    SEX OFFENSES NON FORCIBLE          800 Block of BRYANT ST     12
51              STOLEN PROPERTY          800 Block of BRYANT ST     93
52                      SUICIDE        100 Block of GGBRIDGE HY     21
53               SUSPICIOUS OCC          800 Block of BRYANT ST   1153
60                     TRESPASS        1000 Block of POTRERO AV    248
61                    VANDALISM          800 Block of BRYANT ST   1155
62                VEHICLE THEFT      1500 Block of BAY SHORE BL    162
63                     WARRANTS          800 Block of BRYANT ST   1719
64                  WEAPON LAWS          800 Block of BRYANT ST    156

In [4]:
# The number of crimes an address had the highest frequency for a crime
top_address = address_count.groupby('Address').size().reset_index(0)
top_address

Address   0
0         100 Block of GGBRIDGE HY   1
1         1000 Block of POTRERO AV   3
2            1200 Block of PAGE ST   1
3          1400 Block of PHELPS ST   1
4       1500 Block of BAY SHORE BL   1
5     200 Block of INTERSTATE80 HY   1
6         2000 Block of MISSION ST   1
7           700 Block of KEARNY ST   1
8              800 Block of 3RD ST   1
9           800 Block of BRYANT ST  23
10          800 Block of MARKET ST   1
11              ELLIS ST / HYDE ST   1
12  MISSION ST / SOUTH VAN NESS AV   1

#### Address Break Down
The amount of crime is overwhelmingly clustered around one address. 800 Block of Bryant ST is the most common address for 23 types of crimes, with the second place address being the top for 3. At first glance many of the crimes that occur the most in  Bryant St. are violent crimes or serious property crimes. 

Since the crimes are clustered around one address, using the correlation between address and crime type probably won't be too useful.

Lets try looking at the crime descriptions.

In [5]:
crime_desc = df.groupby(['Category','Descript']).size().reset_index()
crime_desc

Category                                           Descript     0
0          ARSON                                              ARSON   447
1          ARSON                     ARSON OF A COMMERCIAL BUILDING    65
2          ARSON                         ARSON OF A POLICE BUILDING     2
3          ARSON                          ARSON OF A POLICE VEHICLE     1
4          ARSON                         ARSON OF A VACANT BUILDING    33
5          ARSON                                 ARSON OF A VEHICLE   607
6          ARSON                     ARSON OF AN INHABITED DWELLING   217
7          ARSON                                    ATTEMPTED ARSON   109
8          ARSON                           FIRE, UNLAWFULLY CAUSING    32
9        ASSAULT      AGGRAVATED ASSAULT OF POLICE OFFICER, SNIPING     1
10       ASSAULT  AGGRAVATED ASSAULT OF POLICE OFFICER,BODILY FORCE   170
11       ASSAULT    AGGRAVATED ASSAULT ON POLICE OFFICER WITH A GUN     6
12       ASSAULT  AGGRAVATED ASSAULT ON POLICE OFFICER WITH A KNIFE    19
13       ASSAULT            AGGRAVATED ASSAULT WITH A DEADLY WEAPON  6451
14       ASSAULT                      AGGRAVATED ASSAULT WITH A GUN  1071
15       ASSAULT                    AGGRAVATED ASSAULT WITH A KNIFE  2442
16       ASSAULT               AGGRAVATED ASSAULT WITH BODILY FORCE  4898
17       ASSAULT                                            ASSAULT   502
18       ASSAULT                               ASSAULT BY POISONING     7
19       ASSAULT                          ASSAULT BY POLICE OFFICER     7
20       ASSAULT   ASSAULT ON A POLICE OFFICER WITH A DEADLY WEAPON   253
21       ASSAULT    ASSAULT OR ATTEMPTED MURDER UPON GOV'T OFFICERS     1
22       ASSAULT                     ASSAULT WITH CAUSTIC CHEMICALS   399
23       ASSAULT  ASSAULT, AGGRAVATED, ON POLICE OFFICER, W/ FUL...     1
24       ASSAULT     ASSAULT, AGGRAVATED, ON POLICE OFFICER, W/ GUN    10
25       ASSAULT  ASSAULT, AGGRAVATED, ON POLICE OFFICER, W/ SEM...     1
26       ASSAULT                        ASSAULT, AGGRAVATED, W/ GUN   211
27       ASSAULT                ASSAULT, AGGRAVATED, W/ MACHINE GUN     2
28       ASSAULT                  ASSAULT, AGGRAVATED, W/ SEMI AUTO     3
29       ASSAULT         ATTEMPTED HOMICIDE WITH A DANGEROUS WEAPON    63
..           ...                                                ...   ...
857  WEAPON LAWS     FIREARM, LOADED, IN VEHICLE, POSSESSION OR USE   219
858  WEAPON LAWS                       FIREARM, NEGLIGENT DISCHARGE    24
859  WEAPON LAWS          FIREARM, POSSESSION OF WHILE WEARING MASK     5
860  WEAPON LAWS          FLAMMABLE OR EXPLOSIVE DEVICE, POSSESSION    23
861  WEAPON LAWS                ILLEGAL TRANSPORTAION OF EXPLOSIVES     3
862  WEAPON LAWS                           LOCATED EXPLOSIVE DEVICE    80
863  WEAPON LAWS          LOITERING WHILE CARRYING CONCEALED WEAPON   352
864  WEAPON LAWS       POSS OF DEADLY WEAPON WITH INTENT TO ASSAULT   254
865  WEAPON LAWS                                  POSS OF FIRE BOMB    17
866  WEAPON LAWS    POSS OF FIREARM BY CONVICTED FELON/ADDICT/ALIEN   654
867  WEAPON LAWS                           POSS OF FIREARM SILENCER     3
868  WEAPON LAWS                             POSS OF LOADED FIREARM  1963
869  WEAPON LAWS                          POSS OF PROHIBITED WEAPON   648
870  WEAPON LAWS                            POSS OF TEAR GAS WEAPON    53
871  WEAPON LAWS                              POSSESSION OF AIR GUN   408
872  WEAPON LAWS                     POSSESSION OF CAUSTIC CHEMICAL    26
873  WEAPON LAWS                     POSSESSION OF EXPLOSIVE DEVICE    22
874  WEAPON LAWS                          POSSESSION OF MACHINE GUN    11
875  WEAPON LAWS                        SNIPER SCOPE, POSSESSION OF     1
876  WEAPON LAWS                                 SUSPICIOUS PACKAGE    25
877  WEAPON LAWS                      SWITCHBLADE KNIFE, POSSESSION    85
878  WEAPON LAWS                    TAMPERING WITH MARKS ON FIREARM    36
879  WEAPON 

After looking at these results it looks like the description is a strong indicator for the category of the crime. The issue is I need to get my model to see the similarities between each of the differenct descriptions.

#### Oops...
Looks like the descript data is provided only in the training data : / . Well with that let me move on to some modeling.

In [6]:
# We'll be starting off with 
import sklearn.ensemble as sk
import numpy as np
rfc = sk.RandomForestClassifier(oob_score = True, n_estimators = 50, warm_start=True)
Y = df['Category'].astype('category')
Y = Y.cat.rename_categories(range(0,Y.nunique()))

X =  df.drop(['Category','Descript','Resolution'],1)
n = 0
for col in X.columns:
    if n < 4:
        X[col] = X[col].astype('category')
        X[col] = X[col].cat.rename_categories(range(0,X[col].nunique()))
    n += 1

Dates DayOfWeek PdDistrict Address           X          Y
0       389256         6          4   19790 -122.425892  37.774599
1       389256         6          4   19790 -122.425892  37.774599
2       389255         6          4   22697 -122.424363  37.800414
3       389254         6          4    4266 -122.426995  37.800873
4       389254         6          5    1843 -122.438738  37.771541
5       389254         6          2    1505 -122.403252  37.713431
6       389254         6          2   13322 -122.423327  37.725138
7       389254         6          0   18054 -122.371274  37.727564
8       389253         6          6   11384 -122.508194  37.776601
9       389253         6          1   17658 -122.419088  37.807802
10      389252         6          1   17658 -122.419088  37.807802
11      389251         6          8     538 -122.487983  37.737667
12      389251         6          9   22379 -122.412414  37.783004
13      389250         6          4   15935 -122.432915  37.784353
14      389249         6          0    6261 -122.397744  37.729935
15      389249         6          0    1001 -122.383692  37.743189
16      389249         6          9   15549 -122.412597  37.783932
17      389248         6          2   16575 -122.421682  37.742822
18      389247         6          0   19077 -122.386401  37.738983
19      389246         6          9    2208 -122.412250  37.782556
20      389246         6          2    5789 -122.449389  37.742669
21      389245         6          2    4621 -122.420272  37.747332
22      389244         6          9    2208 -122.412250  37.782556
23      389244         6          9    2208 -122.412250  37.782556
24      389243         6          4   15947 -122.436049  37.799841
25      389242         6          9    8671 -122.410509  37.786043
26      389242         6          4    6327 -122.431018  37.787388
27      389242         6          2   10903 -122.423656  37.732556
28      389242         6          8    5454 -122.475773  37.744919
29      389241         6          8    6305 -122.499787  37.748518
...        ...       ...        ...     ...         ...        ...
878019      19         1          7   11808 -122.410294  37.782231
878020      18         1          4   22662 -122.420642  37.781961
878021      17         1          4   21265 -122.431191  37.767595
878022      16         1          3    4970 -122.419516  37.763429
878023      15         1          9   11495 -122.416894  37.784286
878024      14         1          5   11482 -122.437781  37.775483
878025      13         1          0   19605 -122.387710  37.740674
878026      13         1          0   19605 -122.387710  37.740674
878027      12         1          7   12766 -122.409708  37.782828
878028      12         1          7   11808 -122.410294  37.782231
878029      11         1          9    3928 -122.434423  37.779193
878030      11         1          9    3928 -122.434423  37.779193
878031      10         1          0    9752 -122.390417  37.735593
878032       9         1          4   16444 -122.423031  37.785482
878033       8         1          6    2728 -122.391668  37.757793
878034       8         1          6    2728 -122.391668  37.757793
878035       7         1          4    3802 -122.431046  37.783030
878036       7         1          4    3802 -122.431046  37.783030
878037       7         1          4    3802 -122.431046  37.783030
878038       6         1          9   22104 -122.411519  37.786941
878039       5         1          4   20491 -122.420692  37.790577
878040       4         1          3    8010 -122.414073  37.751685
878041       3         1          6   14793 -122.472985  37.782552
878042       2         1          0    4317 -122.389769  37.730564
878043       2         1          0    4317 -122.389769  37.730564
878044       1         1          8   15816 -122.459033  37.714056
878045       0         1          2   11491 -122.447364  37.731948
878046       0         1          7   11315 -122.403390 

In [7]:
# %time model = rfc.fit(X.ix[0:350000],Y.ix[0:350000])
# model.oob_score_

Pretty Much my Random Forest sucks, I'll try making additional features and see how things turn out.

In [8]:
# Breaking down Time stamp into Year, Month, and Day
X['Dates'] = pd.to_datetime(df['Dates'])
X['Year'] = X.Dates.dt.year
X['Month'] = X.Dates.dt.month
X['Day'] = X.Dates.dt.day
X.drop('Dates', 1,inplace=True)
X



DayOfWeek PdDistrict Address           X          Y  Year  Month  Day
0              6          4   19790 -122.425892  37.774599  2015      5   13
1              6          4   19790 -122.425892  37.774599  2015      5   13
2              6          4   22697 -122.424363  37.800414  2015      5   13
3              6          4    4266 -122.426995  37.800873  2015      5   13
4              6          5    1843 -122.438738  37.771541  2015      5   13
5              6          2    1505 -122.403252  37.713431  2015      5   13
6              6          2   13322 -122.423327  37.725138  2015      5   13
7              6          0   18054 -122.371274  37.727564  2015      5   13
8              6          6   11384 -122.508194  37.776601  2015      5   13
9              6          1   17658 -122.419088  37.807802  2015      5   13
10             6          1   17658 -122.419088  37.807802  2015      5   13
11             6          8     538 -122.487983  37.737667  2015      5   13
12             6          9   22379 -122.412414  37.783004  2015      5   13
13             6          4   15935 -122.432915  37.784353  2015      5   13
14             6          0    6261 -122.397744  37.729935  2015      5   13
15             6          0    1001 -122.383692  37.743189  2015      5   13
16             6          9   15549 -122.412597  37.783932  2015      5   13
17             6          2   16575 -122.421682  37.742822  2015      5   13
18             6          0   19077 -122.386401  37.738983  2015      5   13
19             6          9    2208 -122.412250  37.782556  2015      5   13
20             6          2    5789 -122.449389  37.742669  2015      5   13
21             6          2    4621 -122.420272  37.747332  2015      5   13
22             6          9    2208 -122.412250  37.782556  2015      5   13
23             6          9    2208 -122.412250  37.782556  2015      5   13
24             6          4   15947 -122.436049  37.799841  2015      5   13
25             6          9    8671 -122.410509  37.786043  2015      5   13
26             6          4    6327 -122.431018  37.787388  2015      5   13
27             6          2   10903 -122.423656  37.732556  2015      5   13
28             6          8    5454 -122.475773  37.744919  2015      5   13
29             6          8    6305 -122.499787  37.748518  2015      5   13
...          ...        ...     ...         ...        ...   ...    ...  ...
878019         1          7   11808 -122.410294  37.782231  2003      1    6
878020         1          4   22662 -122.420642  37.781961  2003      1    6
878021         1          4   21265 -122.431191  37.767595  2003      1    6
878022         1          3    4970 -122.419516  37.763429  2003      1    6
878023         1          9   11495 -122.416894  37.784286  2003      1    6
878024         1          5   11482 -122.437781  37.775483  2003      1    6
878025         1          0   19605 -122.387710  37.740674  2003      1    6
878026         1          0   19605 -122.387710  37.740674  2003      1    6
878027         1          7   12766 -122.409708  37.782828  2003      1    6
878028         1          7   11808 -122.410294  37.782231  2003      1    6
878029         1          9    3928 -122.434423  37.779193  2003      1    6
878030         1          9    3928 -122.434423  37.779193  2003      1    6
878031         1          0    9752 -122.390417  37.735593  2003      1    6
878032         1          4   16444 -122.423031  37.785482  2003      1    6
878033         1          6    2728 -122.391668  37.757793  2003      1    6
878034         1          6    2728 -122.391668  37.757793  2003      1    6
878035         1          4    3802 -122.431046  37.783030  2003      1    6
878036         1          4    3802 -122.431046  37.783030  2003      1    6
878037         1          4    3802 -122.431046  37.783030  2003      1    6
878038         1          9   22104 -122.411519  37.786941  2003      1    6
878

In [9]:
# RFC Round 2
n = 250
clf = sk.RandomForestClassifier(oob_score = True,n_jobs=-1, n_estimators = n)

# %time clf.fit(X[::10],Y[::10])



In [10]:
clf.oob_score_

AttributeError: 'RandomForestClassifier' object has no attribute 'oob_score_'

In [ ]:
# model.score(X,Y)
# 0.24406610565014025 (depth 7, 250)
# 0.243709633517 500 7
# 0.244111661194 750 7
# 0.243918050132 1000 7
# 0.29118306609312239 250 15


Spent the week trying to get the Random Forest Working to output a decent oob_score_ but came up empty. I tried principal dimensional analysis, breaking apart the model into small pieces, tried different settings. The reason why I kept trying for so long is I couldn't figure out if it was the model itself or if it is just my computers memory limitations was severely inhibiting the model. Either way it doesn't matter, its time to move on and try something else.

In [ ]:
from sklearn.naive_bayes import GaussianNB
gnb = GaussianNB()

In [ ]:
model = gnb.fit(X,Y)
y_pred = model.predict(X)
y_pred

In [ ]:
model.score(X,Y)

Naive bayes also returned a poor result.

In [9]:
from sklearn.svm import SVC
vcl = SVC()
# %time sv_model = vcl.fit(X[::100],Y[::100]) #29.6s



In [ ]:
# %time sv_model.score(X[::10],Y[::10]) # 1min 20s .278
%time sv_model.score(X[::100],Y[::100]) #8.12s .976

In [31]:
rows = np.random.choice(X.index.values, X.shape[0] * .02)
Z = X.ix[rows]
T = Y.ix[rows]

SVM actually shows some promise when choosing every 100 entries once I try every 10 entries the score drops back down to 27%. Lets see what else I can do to clean to improve the score. 

In [37]:
%time vcl.fit(Z,T)

Wall time: 1min 46s


SVC(C=1.0, cache_size=200, class_weight=None, coef0=0.0, degree=3, gamma=0.0,
  kernel='rbf', max_iter=-1, probability=False, random_state=None,
  shrinking=True, tol=0.001, verbose=False)

In [40]:
%time vcl.score(Z,T)

Wall time: 31.8 s


0.22004555808656037

| Sample Size | Columns | Fit Time | Score Time | Score | Comments and Notes                               |
|-------------|---------|----------|------------|-------|--------------------------------------------------|
| 2%          | 8       | 2min 39s | 36.9s      | 95.9% | Used model on sample  may be due to over fitting |
| 2%          | 8       | 1 min 46s| 31.8       | 96.1% | results of model when run right after .fit()     |
| 2%          | 8       | 1 min 46s| 31.8       | 22.0% | Model used on different sample                   |
|             |         |          |            |       |                                                  |
|             |         |          |            |       |                                                  |

In [41]:
test = pd.read_csv('./data/test.csv')
test.drop('Id',axis=1, inplace=True)
n = 0
test['Dates'] = pd.to_datetime(test['Dates'])
test['Year'] = test.Dates.dt.year
test['Month'] = test.Dates.dt.month
test['Day'] = test.Dates.dt.day
test.drop('Dates', 1,inplace=True)

for col in test.columns:
    if n < 3:
        test[col] = test[col].astype('category')
        test[col] = test[col].cat.rename_categories(range(0,test[col].nunique()))
    n += 1